stimpacks/deploy/config.go

```diff
@@ -0,0 +1,352 @@
+		d.log.Fatal("Error parsing deployment config {}", err)
+	}
+
+	d.config.configFilePath = configFile
+
+	d.processConfig()
+
+}
+
+// processConfig ensures that the deployment config is valid
+func (d *Deploy) processConfig() {
+
+	// Set defaults
+	setConfigDefault(&d.config.Deployment.Container.Repo, defaultContainerRepo)
+	setConfigDefault(&d.config.Deployment.Container.Tag, defaultContainerTag)
+	setConfigDefault(&d.config.Deployment.Directory, defaultDeployDirectory)
+	setConfigDefault(&d.config.Deployment.Script, defaultDeployScript)
+
+	// Create our global spec if it doesn't exist so we don't have to keep checking if it exists
+	if d.config.Global.Spec == nil {
+		d.config.Global.Spec = &Spec{}
+	}
+
+	d.config.environmentMap = make(map[string]int)
+	for i, environment := range d.config.Environments {
+
+		// Check to make sure that we don't have multiple environments with the same name
+		if _, ok := d.config.environmentMap[environment.Name]; ok {
+			d.log.Fatal("Error parsing config, duplicate environment name `{}` found", environment.Name)
+		}
+		d.config.environmentMap[environment.Name] = i
+
+		// Ensure there are instances for this environment
+		if len(environment.Instances) <= 0 {
+			d.log.Fatal("No instances found for environment: `{}`", environment.Name)
+		}
+
+		// Create our environment spec if it doesn't exist so we don't have to keep checking if it exists
+		if environment.Spec == nil {
+			environment.Spec = &Spec{}
+		}
+
+		environment.instanceMap = make(map[string]int)
+		for j, instance := range environment.Instances {
+
+			// Check to make sure that we don't have multiple instances with the same name
+			if _, ok := environment.instanceMap[instance.Name]; ok {
+				d.log.Fatal("Error parsing config, duplicate instance name '{}' for environment '{}'", instance.Name, environment.Name)
+			}
+			environment.instanceMap[instance.Name] = j
+
+			// Ensure the instance name does not conflict with the ALL option name.  This is a reserved name for designating a deployment to all instances in an environment via the manual prompt list
+			if strings.ToLower(instance.Name) == strings.ToLower(allOptionPrompt) || strings.ToLower(instance.Name) == strings.ToLower(allOptionCli) {
+				d.log.Fatal("Deployment config cannot have an instance named '{}'. It is a reserved name.", instance.Name)
+			}
+
+			// Create our instance spec if it doesn't exist so we don't have to keep checking if it exists
+			if instance.Spec == nil {
+				instance.Spec = &Spec{}
+			}
+
+			// Merge all of the secrets and environment variables
+			// Instance-level specs take precedence, followed by environment-level then global-level
+			if instance.Spec.Kubernetes.ServiceAccount == "" {
+				if environment.Spec.Kubernetes.ServiceAccount != "" {
+					instance.Spec.Kubernetes.ServiceAccount = environment.Spec.Kubernetes.ServiceAccount
+				} else {
+					instance.Spec.Kubernetes.ServiceAccount = d.config.Global.Spec.Kubernetes.ServiceAccount
```

Comment:
Do we validate that `d.config.Global.Spec.Kubernetes.ServiceAccount` and below `instance.Spec.Kubernetes.Cluster = d.config.Global.Spec.Kubernetes.Cluster`
are populated??  should probably do that before we start assigning them, or maybe bail out if we get here and they are empty

Reply:
so you do *kindof* have a check for `d.config.Global.Spec.Kubernetes.Cluster` at line 196.  I personally think that should be moved to inside these blocks, so you would have:
```golang
			if instance.Spec.Kubernetes.ServiceAccount == "" {
				if environment.Spec.Kubernetes.ServiceAccount != "" {
					instance.Spec.Kubernetes.ServiceAccount = environment.Spec.Kubernetes.ServiceAccount
				} else if d.config.Global.Spec.Kubernetes.ServiceAccount != null {
					instance.Spec.Kubernetes.ServiceAccount = d.config.Global.Spec.Kubernetes.ServiceAccount
				} else {
					d.log.Fatal("No Service Account Set")
				}

			}
			if instance.Spec.Kubernetes.Cluster == "" {
				if environment.Spec.Kubernetes.Cluster != "" {
					instance.Spec.Kubernetes.Cluster = environment.Spec.Kubernetes.Cluster
				} else if d.config.Global.Spec.Kubernetes.Cluster != null {
					instance.Spec.Kubernetes.Cluster = d.config.Global.Spec.Kubernetes.Cluster
				} else {
					d.log.Fatal("No Kube Cluster Set")
				}

			}
```

Then you can remove the if check at line 196

Reply:
Agreed. Fixed